First draft of code insights direction page new format (#4763)
sourcegraph/about: handbook/direction/code-graph/code-insights/index.md

# Code Insights Direction

Code Insights is an upcoming product that lets you track and understand what's in your code and how it changes over time.

This page outlines the vision, strategy, and goals of the Code Insights team.

Quick links:

> - [Code Insights Backlog](https://github.com/orgs/sourcegraph/projects/200/) <!-- Update links with your page -->

Reviewer: Maybe link to the All issues view? This will link to the current iteration, which could be misleading for someone looking for the _Backlog_

Author: Think the kanban view makes the most sense, will do

- Latest demo – **TODO check with marketing if we're allowed to share a demo here**
- [Documentation](https://docs.sourcegraph.com/code_insights)

## Mission, Vision, and Guiding Principles

### Mission

Code Insights helps engineering leaders, managers, and engineers answer questions about how their code is changing or what's in their code that are difficult or impossible to answer otherwise.

We bring **the best practices of big data to challenges of big code**, turning a codebase into a queryable database.

**We help engineering leaders** set and measure quantitative engineering goals, whether they be related to a migration, deprecation, vulnerability, code smell, or high level understanding. We also help engineering leaders understand the shape of their code and how it changes over time. Then we help them tie trends and metrics to _owners_ on their engineering teams.

> **We help individual engineers** measure the day-to-day goals of their teams: how they are progressing on an individual initiative, or how quickly other teams are adopting their code. We also help them get a historical usage view of functions, changes, or other patterns. Happily, we've heard that Code Insights has even helped some engineers get promotions or bonuses, because they use code insights to quantify the impact and spread of their work.

Reviewer: Maybe we should also mention in a sentence here the value of making things that are obviously visible to engineers "down in the trenches" visible to their engineering leaders, by allowing them to create dashboards with visualizations for problems they are seeing and sharing them with their leadership that their management often doesn't notice. Many engineers struggle to get resources dedicated to paying down tech debt, increasing test coverage, etc. Leadership personas tend to look at graphs, but for business KPIs, not aware of the technical costs and state. Especially powerful if we can eventually overlay those two together to show correlations.

Author: good call, will add!

### Code Insights is an enterprise quality professional tool

For Code Insights to be valuable, it must be precise, exact, and performant. When it comes to codebases, there's little value in a measurement tool that is "90%" correct, or "within +/- 5%": our users need exact answers to inform decisions and track goals successfully. Code Insights also has to be performant enough that one can use it to answer questions when the need arises – our users can't wait weeks for their data, nor can they "ration" use of Code Insights among a Sourcegraph instance.

**What this looks like:**

- We don't compromise on precise accuracy; we test thoroughly, often, and at the extremes of possible use
- We often prioritize improvements to speed and database size above more user-facing tweaks

**What this does _not_ look like:**

- Not every Code Insights feature will be "instant" – our developer audience understands the complexity of the tool and forgives some amount of loading/processing time

### Code Insights builds on top of Sourcegraph's other features

> Users do not need to learn a new query language or format to use Code Insights. Nor will Code Insights introduce features that function differently than those "same" features in other areas of the product, confusing users. Code Insights is built atop Sourcegraph Search and will build with Sourcegraph's other features rather than implement its own versions of any functionalit.

Reviewer: ```suggestion
Users do not need to learn a new query language or format to use Code Insights. Nor will Code Insights introduce features that function differently than those "same" features in other areas of the product, confusing users. Code Insights is built atop Sourcegraph Search and will build with Sourcegraph's other features rather than implement its own versions of any functionality.
```

**What this looks like:**

- If a Code Insights feature shares a name or purpose with another Sourcegraph feature, we'll adopt the name and constraints of that feature
- If building a feature that might be reused across the Sourcegraph platform, Code Insights consults with other product engineering teams

**What this does _not_ look like:**

- Code Insights won't avoid introducing net new features if no other Sourcegraph feature has yet built them

## Where we are now

> Code Insights is a [Beta Feature](../../product/beta_and_experimental_feature_labels.md). It entered Beta in August 2021, and will likely remain in Beta into Q4.

Reviewer: ```suggestion
Code Insights is a [Beta Feature](../../product/beta_and_experimental_feature_labels.md). It entered Beta in August 2021, and will likely remain in Beta into FQ4.
```

This means that the Code Insights team is singularly focused on bringing Code Insights into general availability. To reach that goal, Code Insights will be building stability, features, and core product experiences based on the feedback we get from our beta participants.

During the Beta, we are also evaluating pricing and packaging options. It is likely Code Insights will be a standalone feature with its own pricing.

> For what we have learned so far and how we are operationalizing those learnings, see [Strategy](#Strategy-and-plans).

Reviewer: ```suggestion
For what we have learned so far and how we are operationalizing those learnings, see [Strategy](#strategy-and-plans).
```

### Top customer, support, sales and marketing issues

- The biggest issues customers face, and that we're addressing, are the yet-to-be-built product gaps. See the docs page for [Current Limitations of Code Insights](https://docs.sourcegraph.com/code_insights/explanations/current_limitations_of_code_insights) for an up-to-date list.
- Internally, the main task for Code Insights is determining the pricing and packaging

### Competitive landscape

> While there are many devtool products that use some version of the word "Insights," and a handful of others that do "high level information about code," there are **no known direct compeitors to Code Insights**.

Reviewer: ```suggestion
While there are many devtool products that use some version of the word "Insights," and a handful of others that do "high level information about code," there are **no known direct competitors to Code Insights**.
```

There may be one-off instances of companies that have glued together in-house solutions to answer some of the same problems, but we have already validated (thus far) every one of those companies would prefer to use Code Insights.

While Code Insights is in Beta, you can find a detailed breakdown of "similar" products in the original [Product Document for Code Insights](https://docs.google.com/document/d/1d34gCpt_rUOMAun8phcjNsFofGaaA_N_8znmgaugdKw/edit#bookmark=id.pobwla63lsa7). At some moment pre-GA launch, the product and marketing team will create more standard competitive resources based on the GA product features and merge them into this page.

## Detailed Strategy and Plans

### Strategy for FY22H2:

The next six months are focused on building a product that can quickly prove out its product-market fit via successful customer sales.

### Goals

#### FY22Q3

> - OKR 1: We have qualitative/anecdotal evidence that at least 3 customers have a manager/director/VP who is using code insights
> - OKR 2: 50% of customers on 3.31+ make 10 code insights per customer, and have at least 5 weekly unique viewers of insights ([Looker dashboard](https://sourcegraph.looker.com/dashboards-next/209?Latest+Version=))

Reviewer: These are really the key results, not the objectives, right?
```suggestion
- KR 1: We have qualitative/anecdotal evidence that at least 3 customers have a manager/director/VP who is using code insights
- KR 2: 50% of customers on 3.31+ make 10 code insights per customer, and have at least 5 weekly unique viewers of insights ([Looker dashboard](https://sourcegraph.looker.com/dashboards-next/209?Latest+Version=))
```

Author: Sometimes I think it can be read as "Objective's key result" but sure, also works :)

### Themes (and What's next and why)

#### Powerful but Simple

Our biggest pre-beta learning was that a nice GUI and quick flow to creating a user's first Code Insight dramatically improves initial adoption. As we build a product that needs to be complex and configurable enough to serve the varied needs of our customers and fit perfectly to their codebase, we want to maintain the ease of setting up and navigating Code Insights.

We will:

- Continue abstracting complexity (example: views vs tables) unless necessary to expose it
- Build new setup features to make it easier to track a variety of other use cases, like with regex capture groups
- Explore what "quick set up" insights might be possible the first moment a customer uses insights

#### "Tracking" Use Case Needs

A major use case for Code Insights is tracking migrations and similar efforts. There are a number of core features we've learned are valuable to users that we have yet to build.

We will:

- Enable things like custom x-axis time ranges on Insights
- Allow users to edit insights after creation
- Explore additional chart visualizations that are helpful

> - Allow engineering leaders to export insights data to their preferred dashboards tools

Reviewer: This was only listed as a "stretch project" in the Q3 one-pager, while now it uses the strong wording "we will". Is this because this is for FY22H2 and this being a non-stretch goal for FQ4?

Author: That is correct!

We will:

- Explore new types of filters and "drilldowns"
- Build support for clicking into insights to see the code changes behind them

#### Stability and Performance

We're an early Beta product and have a lot of potential growth in both how stable/observable and how performant Code Insights are. Putting effort into both of these areas now will pay major dividends in developing future features and scaling to larger customers.

We will:

- Migrate Insights configurations out of settings files into a proper database
- Continue to optimize our backend to run more insights over more repos in less time

#### Launch Planning

In order to have a successful GA launch, we are also focused on the tasks required to build excitement, storytelling, and sales energy.

We will:

- Be planning for customer case studies and quotes
- Align on pricing and packaging
- Build landing pages and marketing content ahead of the launch

### What we're not working on & why

In order to focus and have a successful initial launch, the Code Insights team is explicitly not working on any of the following until FY23:

> - Code Insights Fully GA on Cloud: due to the multi-tenant nature of cloud, there's additional backend work we'll do in FY23 to make Insights GA on Cloud. Optimizations we plan to make in FY22H2 will help us quickly plan and support Code Insights on Cloud, but we won't have time for that until FY23H1.

Reviewer: ```suggestion
- Code Insights full GA on Cloud: due to the multi-tenant nature of cloud, there's additional backend work we'll do in FY23 to make Insights GA on Cloud. Optimizations we plan to make in FY22H2 will help us quickly plan and support Code Insights on Cloud, but we won't have time for that until FY23H1.
```

# Code Insights Direction

Code Insights is an upcoming product that lets you track and understand what's in your code and how it changes over time.

This page outlines the vision, strategy, and goals of the Code Insights team.

Quick links:

- [Code Insights Backlog](https://github.com/orgs/sourcegraph/projects/200/views/13)

> - Latest demo – **TODO check with marketing if we're allowed to share a demo here**

Reviewer 2: Amie said she prefers not to include a product demo even of we are making it clear we're currently in beta (and we are!).

Author: Got it, will modify, thanks!

Author: ```suggestion
- Latest demo – private since the product is in beta; slack the  #code-insights team or email feedback@sourcegraph.com if you're curious! 
```

**What this does _not_ look like:**

- If engineering leaders want feature C, but feature C would be _harmful_ to the experience of individual developers, we will not build feature C.

### Code Insights layers together the Sourcegraph feature ecosystem

Code Insights naturally sits at the intersection of many other key features, and there are validated needs for, for example: customers to have Insights visualizing what they monitor with Code Monitors, Batch Changes to naturally flow into Code Insights for tracking progress and impact, Precise Code Intelligence that powers code insights for any specific symbol across all repositories.

As Sourcegraph becomes an increasingly powerful developer tool, Code Insights will be vital to tracking and understanding how both Sourcegraph and a company's engineers impact a codebase.

**What this looks like:**

- Code Insights will be extensible and integrate with other Sourcegraph features

**What this does _not_ look like:**

- We won't add integrations just for the sake of integrations, without a validated need and use case

### Code Insights is NOT for "tracking individual developers"

> Code Insights is not built for trying to track an individual's output and whether or not someone is a "good" engineer or "high performing" employee. There are (unfortunately) other tools that exist that have made the mistake of tracking easy-to-find, pointless-to-measure metrics like "number of commits" or "number of lines of code" an individual contributes.

Reviewer 2: Should we soften the language here?

Author: We totally could, I'm open to it. Something like this?
```suggestion
Code Insights is not built for trying to track an individual's output and whether or not someone is a "good" engineer or "high performing" employee. There are other tools that exist that track what (we think) are relatively low-signal basic git metrics like like "number of commits" or "number of lines of code" an individual contributes.
```

Author: @thisisvic let me know what you're thinking – would love to get this merged in on Wednesday if possible!
